File: netlify/functions/mock_payment_server.js
const { createHmac, randomUUID } = require("crypto");

const CORS_HEADERS = {
  "Access-Control-Allow-Origin": "*",
  "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
  "Access-Control-Allow-Headers": "Authorization, partnerToken, requestId, Content-Type",
};

const JWT_SECRET = process.env.MOCK_AUTH_SECRET || "dev-secret-change-me";
const JWT_ALGORITHM = "HS256";

function jsonResponse(statusCode, payload) {
  return {
    statusCode,
    headers: { "Content-Type": "application/json", ...CORS_HEADERS },
    body: JSON.stringify(payload),
  };
}

function errorResponse(statusCode, code, message) {
  return jsonResponse(statusCode, { errors: [{ code, message }] });
}

function base64UrlEncode(input) {
  return Buffer.from(input)
    .toString("base64")
    .replace(/=/g, "")
    .replace(/\+/g, "-")
    .replace(/\//g, "_");
}

function base64UrlDecode(input) {
  const padded = input.replace(/-/g, "+").replace(/_/g, "/") + "===".slice((input.length + 3) % 4);
  return Buffer.from(padded, "base64").toString("utf8");
}

function createJwt(sub, expiresInSeconds = 3600) {
  const now = Math.floor(Date.now() / 1000);
  const header = { alg: JWT_ALGORITHM, typ: "JWT" };
  const payload = {
    sub,
    iat: now,
    exp: now + expiresInSeconds,
  };
  const encodedHeader = base64UrlEncode(JSON.stringify(header));
  const encodedPayload = base64UrlEncode(JSON.stringify(payload));
  const signingInput = `${encodedHeader}.${encodedPayload}`;
  const signature = createHmac("sha256", JWT_SECRET).update(signingInput).digest("base64");
  const encodedSignature = signature.replace(/=/g, "").replace(/\+/g, "-").replace(/\//g, "_");
  return `${signingInput}.${encodedSignature}`;
}

function verifyJwt(token) {
  const parts = String(token || "").split(".");
  if (parts.length !== 3) {
    return { valid: false, error: errorResponse(401, "INVALID_TOKEN", "Access token is invalid") };
  }

  const [encodedHeader, encodedPayload, encodedSignature] = parts;
  let header;
  let payload;

  try {
    header = JSON.parse(base64UrlDecode(encodedHeader));
    payload = JSON.parse(base64UrlDecode(encodedPayload));
  } catch {
    return { valid: false, error: errorResponse(401, "INVALID_TOKEN", "Access token is invalid") };
  }

  if (header.alg !== JWT_ALGORITHM) {
    return { valid: false, error: errorResponse(401, "INVALID_TOKEN", "Access token is invalid") };
  }

  const signingInput = `${encodedHeader}.${encodedPayload}`;
  const expectedSignature = createHmac("sha256", JWT_SECRET).update(signingInput).digest("base64")
    .replace(/=/g, "")
    .replace(/\+/g, "-")
    .replace(/\//g, "_");

  if (encodedSignature !== expectedSignature) {
    return { valid: false, error: errorResponse(401, "INVALID_TOKEN", "Access token is invalid") };
  }

  const now = Math.floor(Date.now() / 1000);
  if (typeof payload.exp === "number" && payload.exp < now) {
    return { valid: false, error: errorResponse(401, "TOKEN_EXPIRED", "Access token has expired") };
  }

  return { valid: true, payload };
}

function isValidUUID4(value) {
  return /^[0-9a-f]{8}-[0-9a-f]{4}-4[0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(value);
}

function getHeader(headers, name) {
  if (!headers) return null;
  const lower = name.toLowerCase();
  const key = Object.keys(headers).find((k) => k.toLowerCase() === lower);
  return key ? headers[key] : null;
}

function parseBody(event) {
  if (!event.body) return {};

  let body = event.body;
  if (event.isBase64Encoded) {
    body = Buffer.from(body, "base64").toString("utf8");
  }

  const headers = event.headers || {};
  const contentType = String(getHeader(headers, "content-type") || "").toLowerCase();

  if (contentType.includes("application/x-www-form-urlencoded")) {
    const params = new URLSearchParams(body);
    return Object.fromEntries(params.entries());
  }

  if (typeof body === "object") {
    return body;
  }

  try {
    return JSON.parse(body);
  } catch {
    return {};
  }
}

function validateHeaders(partnerToken, requestId) {
  if (!partnerToken || !isValidUUID4(partnerToken)) {
    return errorResponse(400, "MISSING_REQUIRED_HEADER", "partnerToken header is required and must be a valid UUID");
  }
  if (requestId !== null && requestId !== undefined && !isValidUUID4(requestId)) {
    return errorResponse(400, "INVALID_HEADER", "requestId header must be a valid UUID");
  }
  return null;
}

function validateAuth(partnerToken, authorization, requestId) {
  if (authorization) {
    if (!String(authorization).toLowerCase().startsWith("bearer ")) {
      return errorResponse(401, "INVALID_TOKEN", "Authorization header must be a Bearer token");
    }
    const token = String(authorization).split(/\s+/, 2)[1];
    const verification = verifyJwt(token);
    if (!verification.valid) return verification.error;
    if (requestId !== null && requestId !== undefined && !isValidUUID4(requestId)) {
      return errorResponse(400, "INVALID_HEADER", "requestId header must be a valid UUID");
    }
    return null;
  }

  return validateHeaders(partnerToken, requestId);
}

const CUSTOMERS_RESPONSE = [
  {
    customerId: 27573,
    phone: "+1-8973452344",
    city: "New York",
    address2: "",
    postalCode: "10022",
    state: "NY",
    address: "9011 1234 Block",
    lastName: "TestOFLHM",
    firstName: "Automation",
    country: "US",
    email: "DPPAutomationTestUser@gmail.com",
    shippingAddress: {
      state: "", address: "", lastName: "", firstName: "",
      email: "test@gmail.com", country: "", phone: "1234567890",
      city: "", address2: "", postalCode: "",
    },
    vaults: [
      {
        vaultCreated: "2024-04-22T00:20:44.757",
        accNickName: "Visa-Debit",
        isActive: true,
        paymentMethodId: "474a4400-9114-4950-957d-4da198c307ed",
        cardType: "Mastercard",
        maskedPan: "555555******4444",
        expiry: "12/35",
        token: "2632418387214444",
        accountNumber: "",
        routingNumber: "",
        accountType: "Checking",
        billingAddress: {
          firstName: "Jane", lastName: "Doe", address: "123 Main St",
          address2: "Apt 5", city: "San Francisco", state: "CA",
          country: "USA", postalCode: "94111", phone: "650-555-1234",
          email: "jane@email.com",
        },
        subscription: [
          {
            frequency: null, amount: "0", paymentDay: null,
            endDate: "0001-01-01T00:00:00", paymentMonth: null,
            startDate: "0001-01-01T00:00:00", active: false,
            subscriptionId: "00000000-0000-0000-0000-000000000000",
          },
        ],
      },
    ],
  },
  {
    customerId: 27583,
    phone: "+1-8973452344",
    city: "New York",
    address2: "",
    postalCode: "10022",
    state: "NY",
    address: "1922 1234 Block",
    lastName: "TestLEEGY",
    firstName: "Automation",
    country: "US",
    email: "DPPAutomationTestUser@gmail.com",
    shippingAddress: {
      state: "", address: "", lastName: "", firstName: "",
      email: "test@gmail.com", country: "", phone: "1234567890",
      city: "", address2: "", postalCode: "",
    },
    vaults: [],
  },
];

const DEVICES_RESPONSE = {
  devices: [
    {
      deviceId: "4",
      terminalName: "XUAT DESK3500",
      deviceDisplayName: "QA D3500 GD",
      deviceSerialNumber: "221197303251060424702233",
      terminalStatus: "Online",
      multiMIDSharing: "Enabled",
      cloudStatus: "Active",
      sharedMerchantIds: ["6289980008342010"],
      primaryMerchantId: "6289980008342008",
    },
    {
      deviceId: "5",
      terminalName: "XUAT DESK3500",
      deviceDisplayName: "QA D3500",
      deviceSerialNumber: "221197303251060424702233",
      terminalStatus: "Online",
      multiMIDSharing: "Disabled",
      cloudStatus: "Active",
      sharedMerchantIds: [],
      primaryMerchantId: "6289980008342008",
    },
    {
      deviceId: "9",
      terminalName: "QA D3500 WIFI",
      deviceDisplayName: "QA D3500 WIFI",
      deviceSerialNumber: "24702233",
      terminalStatus: "Offline",
      multiMIDSharing: "Enabled",
      cloudStatus: "Inactive",
      sharedMerchantIds: ["6289980008342008"],
      primaryMerchantId: "6289980008342010",
    },
  ],
};

function handlePayments(body, requestId) {
  const { paymentType, amount } = body;

  if (!paymentType) {
    return errorResponse(400, "MISSING_REQUIRED_FIELD", "paymentType is required");
  }
  if (!["Sale", "Authorization", "Recurring"].includes(paymentType)) {
    return errorResponse(400, "INVALID_PAYMENT_TYPE", "paymentType must be one of: Sale, Authorization, Recurring");
  }
  if (!amount || typeof amount !== "object" || !("amount" in amount)) {
    return errorResponse(400, "MISSING_REQUIRED_FIELD", "amount is required");
  }

  const amountValue = parseFloat(amount.amount);
  if (isNaN(amountValue)) {
    return errorResponse(400, "INVALID_AMOUNT", "amount.amount must be a number");
  }
  if (amountValue <= 0) {
    return errorResponse(400, "INVALID_AMOUNT", "amount.amount must be greater than 0");
  }

  const approved = amountValue <= 9999;
  return jsonResponse(200, {
    isPartial: false,
    orderId: "Order123",
    customerId: 4321,
    batchNumber: 131001,
    subscriptionId: "9a2cb7fe-119c-48ef-973e-8299246df7c2",
    fee: { feeAuthResponse: "APPROVED", feeAmount: 1.99 },
    token: "1556778677451110",
    accountResponseData: { avs: "Y", cvv: "Y" },
    amountApproved: amountValue,
    authResponse: approved ? "APPROVED" : "DECLINED",
    responseCode: approved ? 0 : 51,
    responseMessage: null,
    paymentId: "d290f1ee-6c54-4b01-90e6-d701748f0851",
    requestId: requestId || randomUUID(),
    customData: [],
  });
}

exports.handler = async (event) => {
  const method = (event.httpMethod || "GET").toUpperCase();

  if (method === "OPTIONS") {
    return { statusCode: 200, headers: CORS_HEADERS, body: "" };
  }

  const path = event.path || "/";
  const headers = event.headers || {};
  const partnerToken = getHeader(headers, "partnerToken");
  const requestId = getHeader(headers, "requestId");
  const authorization = getHeader(headers, "Authorization");

  if (method === "POST" && path === "/dpp/v1/auth/token") {
    const body = parseBody(event);
    const clientId = body.clientId || body.client_id;
    const clientSecret = body.clientSecret || body.client_secret;

    if (clientId !== "demo-client" || clientSecret !== "demo-secret") {
      return errorResponse(401, "INVALID_CLIENT", "Invalid client credentials");
    }

    return jsonResponse(200, {
      access_token: createJwt(clientId),
      token_type: "bearer",
      expires_in: 3600,
    });
  }

  const authError = validateAuth(partnerToken, authorization, requestId);
  if (authError) return authError;

  if (method === "GET" && path === "/dpp/v1/customers") {
    return jsonResponse(200, CUSTOMERS_RESPONSE);
  }

  if (method === "GET" && path === "/dpp/v1/emv/devices") {
    return jsonResponse(200, DEVICES_RESPONSE);
  }

  if (method === "POST" && path === "/dpp/v1/payments") {
    const body = parseBody(event);
    return handlePayments(body, requestId);
  }

  return errorResponse(404, "ROUTE_NOT_FOUND", "Route not found");
};

File: netlify/functions/mock_payment_server.py
from __future__ import annotations

import base64
import json
import os
from typing import Any
from uuid import UUID, uuid4
from datetime import datetime, timedelta

import jwt

_CORS_HEADERS = {
    "Access-Control-Allow-Origin": "*",
    "Access-Control-Allow-Methods": "GET, POST, OPTIONS",
    "Access-Control-Allow-Headers": "Authorization, partnerToken, requestId, Content-Type",
}

# JWT secret (override with env var in production)
_JWT_SECRET = os.getenv("MOCK_AUTH_SECRET", "dev-secret-change-me")
_JWT_ALGORITHM = "HS256"


def _json_response(status_code: int, payload: dict[str, Any]) -> dict[str, Any]:
    return {
        "statusCode": status_code,
        "headers": {"Content-Type": "application/json", **_CORS_HEADERS},
        "body": json.dumps(payload),
        "isBase64Encoded": False,
    }


def _error_response(status_code: int, code: str, message: str) -> dict[str, Any]:
    return _json_response(status_code, {"errors": [{"code": code, "message": message}]})


class APIError(Exception):
    def __init__(self, status_code: int, code: str, message: str) -> None:
        self.status_code = status_code
        self.code = code
        self.message = message
        super().__init__(message)


def _is_valid_uuid4(value: str) -> bool:
    try:
        parsed = UUID(value)
    except ValueError:
        return False
    return parsed.version == 4


def _validate_headers(partner_token: str | None, request_id: str | None) -> str | None:
    if not partner_token or not _is_valid_uuid4(partner_token):
        raise APIError(
            400,
            "MISSING_REQUIRED_HEADER",
            "partnerToken header is required and must be a valid UUID",
        )
    if request_id is not None and not _is_valid_uuid4(request_id):
        raise APIError(
            400,
            "INVALID_HEADER",
            "requestId header must be a valid UUID",
        )
    return request_id


CUSTOMERS_RESPONSE = [
    {
        "customerId": 27573,
        "phone": "+1-8973452344",
        "city": "New York",
        "address2": "",
        "postalCode": "10022",
        "state": "NY",
        "address": "9011 1234 Block",
        "lastName": "TestOFLHM",
        "firstName": "Automation",
        "country": "US",
        "email": "DPPAutomationTestUser@gmail.com",
        "shippingAddress": {
            "state": "",
            "address": "",
            "lastName": "",
            "firstName": "",
            "email": "test@gmail.com",
            "country": "",
            "phone": "1234567890",
            "city": "",
            "address2": "",
            "postalCode": "",
        },
        "vaults": [
            {
                "vaultCreated": "2024-04-22T00:20:44.757",
                "accNickName": "Visa-Debit",
                "isActive": True,
                "paymentMethodId": "474a4400-9114-4950-957d-4da198c307ed",
                "cardType": "Mastercard",
                "maskedPan": "555555******4444",
                "expiry": "12/35",
                "token": "2632418387214444",
                "accountNumber": "",
                "routingNumber": "",
                "accountType": "Checking",
                "billingAddress": {
                    "firstName": "Jane",
                    "lastName": "Doe",
                    "address": "123 Main St",
                    "address2": "Apt 5",
                    "city": "San Francisco",
                    "state": "CA",
                    "country": "USA",
                    "postalCode": "94111",
                    "phone": "650-555-1234",
                    "email": "jane@email.com",
                },
                "subscription": [
                    {
                        "frequency": None,
                        "amount": "0",
                        "paymentDay": None,
                        "endDate": "0001-01-01T00:00:00",
                        "paymentMonth": None,
                        "startDate": "0001-01-01T00:00:00",
                        "active": False,
                        "subscriptionId": "00000000-0000-0000-0000-000000000000",
                    }
                ],
            }
        ],
    },
    {
        "customerId": 27583,
        "phone": "+1-8973452344",
        "city": "New York",
        "address2": "",
        "postalCode": "10022",
        "state": "NY",
        "address": "1922 1234 Block",
        "lastName": "TestLEEGY",
        "firstName": "Automation",
        "country": "US",
        "email": "DPPAutomationTestUser@gmail.com",
        "shippingAddress": {
            "state": "",
            "address": "",
            "lastName": "",
            "firstName": "",
            "email": "test@gmail.com",
            "country": "",
            "phone": "1234567890",
            "city": "",
            "address2": "",
            "postalCode": "",
        },
        "vaults": [],
    },
]

DEVICES_RESPONSE = {
    "devices": [
        {
            "deviceId": "4",
            "terminalName": "XUAT DESK3500",
            "deviceDisplayName": "QA D3500 GD",
            "deviceSerialNumber": "221197303251060424702233",
            "terminalStatus": "Online",
            "multiMIDSharing": "Enabled",
            "cloudStatus": "Active",
            "sharedMerchantIds": ["6289980008342010"],
            "primaryMerchantId": "6289980008342008",
        },
        {
            "deviceId": "5",
            "terminalName": "XUAT DESK3500",
            "deviceDisplayName": "QA D3500",
            "deviceSerialNumber": "221197303251060424702233",
            "terminalStatus": "Online",
            "multiMIDSharing": "Disabled",
            "cloudStatus": "Active",
            "sharedMerchantIds": [],
            "primaryMerchantId": "6289980008342008",
        },
        {
            "deviceId": "9",
            "terminalName": "QA D3500 WIFI",
            "deviceDisplayName": "QA D3500 WIFI",
            "deviceSerialNumber": "24702233",
            "terminalStatus": "Offline",
            "multiMIDSharing": "Enabled",
            "cloudStatus": "Inactive",
            "sharedMerchantIds": ["6289980008342008"],
            "primaryMerchantId": "6289980008342010",
        },
    ]
}


def _header_value(headers: dict[str, Any] | None, name: str) -> str | None:
    if not headers:
        return None
    lower_name = name.lower()
    for key, value in headers.items():
        if str(key).lower() == lower_name:
            return value
    return None


def _parse_body(event: dict[str, Any]) -> dict[str, Any]:
    body = event.get("body")
    if body is None or body == "":
        return {}
    if event.get("isBase64Encoded"):
        body = base64.b64decode(body).decode("utf-8")
    # If content-type is form encoded, parse into dict
    headers = event.get("headers") or {}
    content_type = _header_value(headers, "content-type") or ""
    if isinstance(body, dict):
        return body
    if "application/x-www-form-urlencoded" in content_type:
        try:
            from urllib.parse import parse_qs

            parsed = parse_qs(body)
            # parse_qs returns lists for each value
            return {k: v[0] if isinstance(v, list) and v else v for k, v in parsed.items()}
        except Exception:
            return {}
    try:
        return json.loads(body)
    except Exception:
        return {}


def _create_payment_response(payload: dict[str, Any], request_id: str | None) -> dict[str, Any]:
    payment_type = payload.get("paymentType")
    if payment_type is None:
        raise APIError(400, "MISSING_REQUIRED_FIELD", "paymentType is required")
    if payment_type not in {"Sale", "Authorization", "Recurring"}:
        raise APIError(
            400,
            "INVALID_PAYMENT_TYPE",
            "paymentType must be one of: Sale, Authorization, Recurring",
        )

    amount_obj = payload.get("amount")
    if not isinstance(amount_obj, dict) or "amount" not in amount_obj:
        raise APIError(400, "MISSING_REQUIRED_FIELD", "amount is required")

    try:
        amount_value = float(amount_obj["amount"])
    except (TypeError, ValueError):
        raise APIError(400, "INVALID_AMOUNT", "amount.amount must be a number")

    if amount_value <= 0:
        raise APIError(400, "INVALID_AMOUNT", "amount.amount must be greater than 0")

    approved = amount_value <= 9999
    return {
        "isPartial": False,
        "orderId": "Order123",
        "customerId": 4321,
        "batchNumber": 131001,
        "subscriptionId": "9a2cb7fe-119c-48ef-973e-8299246df7c2",
        "fee": {"feeAuthResponse": "APPROVED", "feeAmount": 1.99},
        "token": "1556778677451110",
        "accountResponseData": {"avs": "Y", "cvv": "Y"},
        "amountApproved": amount_value,
        "authResponse": "APPROVED" if approved else "DECLINED",
        "responseCode": 0 if approved else 51,
        "responseMessage": None,
        "paymentId": "d290f1ee-6c54-4b01-90e6-d701748f0851",
        "requestId": request_id or str(uuid4()),
        "customData": [],
    }

def _create_jwt(sub: str, expires_in: int = 3600) -> str:
    now = datetime.utcnow()
    payload = {
        "sub": sub,
        "iat": int(now.timestamp()),
        "exp": int((now + timedelta(seconds=expires_in)).timestamp()),
    }
    return jwt.encode(payload, _JWT_SECRET, algorithm=_JWT_ALGORITHM)

def _verify_jwt(token: str) -> dict | None:
    try:
        return jwt.decode(token, _JWT_SECRET, algorithms=[_JWT_ALGORITHM])
    except jwt.ExpiredSignatureError:
        return {"error": _error_response(401, "TOKEN_EXPIRED", "Access token has expired")}
    except jwt.InvalidTokenError:
        return {"error": _error_response(401, "INVALID_TOKEN", "Access token is invalid")}


def handler(event: dict[str, Any], context: Any) -> dict[str, Any]:
    del context

    path = event.get("path") or event.get("rawPath") or "/"
    method = str(
        event.get("httpMethod")
        or event.get("requestContext", {}).get("http", {}).get("method", "GET")
    ).upper()
    headers = event.get("headers")
    partner_token = _header_value(headers, "partnerToken")
    request_id = _header_value(headers, "requestId")
    authorization = _header_value(headers, "Authorization")

    if method == "OPTIONS":
        return {"statusCode": 200, "headers": _CORS_HEADERS, "body": ""}

    try:
        # If Authorization Bearer token is present, validate it. Otherwise require partnerToken header.
        if authorization:
            if not str(authorization).lower().startswith("bearer "):
                return _error_response(401, "INVALID_TOKEN", "Authorization header must be a Bearer token")
            token = str(authorization).split(None, 1)[1]
            verify_result = _verify_jwt(token)
            if isinstance(verify_result, dict) and verify_result.get("error"):
                return verify_result["error"]
        else:
            _validate_headers(partner_token, request_id)

        # Token issuance endpoint (client credentials demo)
        if method == "POST" and path == "/dpp/v1/auth/token":
            try:
                body = _parse_body(event)
            except json.JSONDecodeError:
                return _error_response(400, "INVALID_REQUEST", "Request body is invalid")
            client_id = body.get("clientId") or body.get("client_id")
            client_secret = body.get("clientSecret") or body.get("client_secret")
            if client_id != "demo-client" or client_secret != "demo-secret":
                return _error_response(401, "INVALID_CLIENT", "Invalid client credentials")
            token = _create_jwt(client_id)
            return _json_response(200, {"access_token": token, "token_type": "bearer", "expires_in": 3600})

        if method == "GET" and path == "/dpp/v1/customers":
            return _json_response(200, CUSTOMERS_RESPONSE)
        if method == "GET" and path == "/dpp/v1/emv/devices":
            return _json_response(200, DEVICES_RESPONSE)
        if method == "POST" and path == "/dpp/v1/payments":
            payload = _parse_body(event)
            return _json_response(200, _create_payment_response(payload, request_id))
    except APIError as exc:
        return _error_response(exc.status_code, exc.code, exc.message)
    except json.JSONDecodeError:
        return _error_response(400, "INVALID_REQUEST", "Request body is invalid")

    return _error_response(404, "ROUTE_NOT_FOUND", "Route not found")